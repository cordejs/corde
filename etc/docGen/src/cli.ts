#!/usr/bin/env node

import { program } from 'commander';
import { mkdirSync, writeFileSync } from 'fs';
import { basename, dirname, isAbsolute, resolve } from 'path';
import {
  CompilerOptions,
  getParsedCommandLineOfConfigFile,
  sys,
} from 'typescript';
import { createDocumentation, Options } from '.';
import { heading, joinSections } from './markdown';
import { formatDiagnosticError } from './utils';

type CLIOptions = {
  project: string;
  entry: string;
  output: string;
};

program
  .name('typescript-documentation')
  .description(
    'Generate markdown API documentation directly from TypeScript source code'
  )
  .option(
    '-p, --project <tsconfig file>',
    'relative or absolute path to a tsconfig.json file',
    './tsconfig.json'
  )
  .option(
    '-e, --entry <main file>',
    'entry/main file of project',
    './src/index.ts'
  )
  .option(
    '-o, --output <markdown file>',
    'markdown documentation output file location',
    './docs/README.md'
  );

function getCompilerOptions(cliOptions: CLIOptions): CompilerOptions {
  const tsConfigPath = isAbsolute(cliOptions.project)
    ? cliOptions.project
    : resolve(process.cwd(), cliOptions.project);

  const config = getParsedCommandLineOfConfigFile(
    tsConfigPath,
    {},
    {
      ...sys,
      onUnRecoverableConfigFileDiagnostic: /* istanbul ignore next */ (
        diagnostic
      ) => {
        /* istanbul ignore next */
        throw new Error(formatDiagnosticError(diagnostic));
      },
    }
  );

  /* istanbul ignore next */
  if (!config) {
    throw new Error(`Unable to parse ${tsConfigPath}`);
  }
  return config.options;
}

function getOutput(cliOptions: CLIOptions): string {
  return isAbsolute(cliOptions.output)
    ? cliOptions.output
    : resolve(process.cwd(), cliOptions.output);
}

function getOptions(cliOptions: CLIOptions): Options {
  try {
    mkdirSync(dirname(getOutput(cliOptions)));
  } catch {
    /* istanbul ignore next */
    1;
  }

  return {
    compilerOptions: getCompilerOptions(cliOptions),
    entry: isAbsolute(cliOptions.entry)
      ? cliOptions.entry
      : resolve(process.cwd(), cliOptions.entry),
    getSectionLocation: (section: string): string =>
      section === 'default'
        ? basename(getOutput(cliOptions))
        : `${section.toLowerCase().replace(/ /g, '-')}.md`,
  };
}

program.parse(process.argv);
const cliOptions: CLIOptions = program.opts();
const options = getOptions(cliOptions);
try {
  createDocumentation(options).forEach((text: string, section: string) => {
    const content =
      section === 'default' ? text : joinSections([heading(section, 1), text]);

    /* istanbul ignore next */
    if (!content) {
      return;
    }

    writeFileSync(
      resolve(
        dirname(getOutput(cliOptions)),
        options.getSectionLocation(section)
      ),
      content,
      'utf8'
    );
  });
} catch (e) {
  /* istanbul ignore next */
  if (e instanceof Error) {
    console.log(e.stack);
  }
  /* istanbul ignore next */
  process.exit(1);
}
