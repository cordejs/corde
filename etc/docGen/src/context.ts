import { Symbol, TypeChecker } from 'typescript';

export type RenderContext = {
  typeChecker: TypeChecker;
  exportedSymbols: Symbol[];
  section: string;
  getSectionLocation: (section: string) => string;
};

export type DependencyContext = {
  typeChecker: TypeChecker;
  exportedSymbols: Symbol[];
  resolutionPath: Symbol[];
};
