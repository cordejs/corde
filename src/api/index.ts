export * from "./botAPI";
export * from "./configAPI";
