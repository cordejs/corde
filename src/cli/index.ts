/* eslint-disable no-console */

process.stdin.resume();

import { Command } from "commander";
import { pack } from "../package";
import { initEnvVariables } from "../envVariables";
import { initEventHandlers } from "../eventHandlers";
import runtime from "../core/runtime";
import { injectGlobals } from "../core/injectGlobals";
import { logger } from "../core/Logger";
import { commandFactory } from "./common";

export const program = new Command();

try {
  initEventHandlers();
  initEnvVariables();

  // global variables can not be injected when running unity tests
  // to not conflict with jest

  if (!runtime.isUnityTest) {
    logger.mock();
    injectGlobals();
  }

  // Add basic information with default run all command
  program
    .name("corde")
    .usage("to start testings o corde [option] to use a specific command.")
    .description(pack.description)
    .version(`v${pack.version}`, "-v, --version");

  commandFactory.loadCommands(program);
} catch (error) {
  logger.log(error);
}

if (process.env.ENV !== "UNITY_TEST" && process.env.ENV !== "E2E_TEST") {
  _main();
}

export async function _main(args?: string[]) {
  await program.parseAsync(args ?? process.argv);
  logger.printStacks();
}
