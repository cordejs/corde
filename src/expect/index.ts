/* eslint-disable no-console */
import chalk from "chalk";
import { runtime } from "../common/runtime";
import { ITestProps, ITestReport } from "../types";
import { corde } from "../types/globals";
import { getStackTrace, typeOf } from "../utils";
import { any } from "./asymmetricMatcher";
import * as matchers from "./matchers";

interface IMatcher {
  (props: ITestProps, ...args: any[]): ITestReport;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function createMatcherFn(matcher: string, isNot: boolean, expected: any, isDebug: boolean) {
  return (...args: any[]): ITestReport | void => {
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    if (expected === any) {
      args = [expected(), ...args];
    } else {
      args = [expected, ...args];
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const trace = getStackTrace(undefined, true, matcher);
    try {
      const matcherFn = pickFn(matcher as KeyOfMatcher);

      const props: ITestProps = {
        isNot,
        expectedColorFn: chalk.bold,
        receivedColorFn: chalk.bold,
        createHint: (...paramsName: string[]) => {
          return `expect(${chalk.green("expected")}).${isNot ? "not." : ""}${matcher}(${paramsName
            .map((param) => chalk.red(param))
            .join(",")})`;
        },
        formatValue: (value: any) => {
          if (typeof value === "symbol") {
            return value.toString();
          }

          if (typeOf(value) === "object") {
            if (Object.keys(value).length === 0) {
              return "{}";
            }

            return "{ ... }";
          }

          if (
            typeof value === "string" &&
            (value.length === 0 || (!value.startsWith("'") && !value.endsWith("'")))
          ) {
            return `'${value}'`;
          }

          if (Array.isArray(value)) {
            if (value.length > 0) {
              return "[...]";
            }
            return "[]";
          }

          return value;
        },
      };

      const report = matcherFn.bind(props, ...args)();
      if (!report.pass) {
        report.trace = trace;
      }

      runtime.internalEvents.emit("test_end", report);

      if (isDebug) {
        return report;
      }
    } catch (error) {
      const failedReport: ITestReport = {
        pass: false,
        message: handleError(error),
        trace: trace,
      };
      runtime.internalEvents.emit("test_end", failedReport);
      if (isDebug) {
        return failedReport;
      }
    }
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message;
  }
  return error;
}

function createLocalExpect(isDebug: boolean) {
  let localExpect: any = {};

  localExpect = (expected: any) => {
    const _expect: any = {};
    _expect.not = {};
    Object.getOwnPropertyNames(matchers).forEach((matcher) => {
      _expect[matcher] = createMatcherFn(matcher, false, expected, isDebug);
      _expect.not[matcher] = createMatcherFn(matcher, true, expected, isDebug);
    });
    return _expect;
  };

  localExpect.any = any;

  return localExpect;
}

export const expect = createLocalExpect(false) as corde.IExpect;

type DebugExpectType<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => any
    ? (...params: Parameters<T[P]>) => { pass: boolean; message: string }
    : DebugExpectType<T[P]>;
};

export interface IDebugExpect {
  <T extends any>(value: T): DebugExpectType<corde.IMatchersWithNot<T>>;
  any(...classType: any[]): any;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const cordeExpect = createLocalExpect(true) as IDebugExpect;
