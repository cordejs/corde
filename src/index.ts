export { group, test, group as describe, test as it } from "./closures";
export { expect } from "./expect";
export { afterAll, beforeStart, afterEach, beforeEach } from "./hooks";
export * from "./types/general";

import { corde } from "./corde";
export const bot = corde.bot;

export default corde;
