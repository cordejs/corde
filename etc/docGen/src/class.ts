import { JSDocTagInfo, Symbol, SymbolFlags } from 'typescript';
import { renderAdditionalLinks } from './additionalLinks';
import { DependencyContext, RenderContext } from './context';
import { renderDescription } from './description';
import { renderExamples } from './examples';
import { heading, joinSections } from './markdown';
import { getSymbolDependencies } from './symbol';
import { getSymbolSection, isInternalSymbol } from './utils';

export function getClassDependencies(
  symbol: Symbol,
  context: DependencyContext
): Symbol[] {
  const members: Symbol[] = [];

  /* istanbul ignore else */
  if (symbol.members) {
    symbol.members.forEach((member) => {
      if (!isInternalSymbol(member)) {
        members.push(member);
      }
    });
  }

  return members.reduce<Symbol[]>(
    (dependencies, member) => [
      ...dependencies,
      ...getSymbolDependencies(member, context),
    ],
    []
  );
}

export function spreadClassProperties(
  symbols: Symbol[],
  getSectionLocation: (section: string) => string
): Symbol[] {
  return symbols.reduce<Symbol[]>((acc, symbol) => {
    if (!(symbol.getFlags() & SymbolFlags.Class) || !symbol.members) {
      return [...acc, symbol];
    }

    const classInstanceName = [
      symbol.name.charAt(0).toLowerCase(),
      symbol.name.slice(1),
    ].join('');
    const section = getSymbolSection(symbol);
    const members: Symbol[] = [];
    const memberSections = new Set<string>();
    symbol.members.forEach((member) => {
      if (!isInternalSymbol(member)) {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        member.getName = (): string => `${classInstanceName}.${member.name}`;
        members.push(member);
        const memberSection = getSymbolSection(member);
        if (memberSection !== 'default' && memberSection !== section) {
          memberSections.add(getSymbolSection(member));
        }
      }
    });

    const originalTags = symbol.getJsDocTags();
    const memberSectionsArray = Array.from<string>(memberSections.values());
    memberSectionsArray.sort();
    const additionalMemberReferences = memberSectionsArray.map<JSDocTagInfo>(
      (section: string) => ({
        name: 'see',
        text: [
          {
            text: `{@link ${getSectionLocation(section)}|${section}}`,
            kind: 'text',
          },
        ],
      })
    );
    // eslint-disable-next-line @typescript-eslint/unbound-method
    symbol.getJsDocTags = (): JSDocTagInfo[] => {
      return [...originalTags, ...additionalMemberReferences];
    };

    return [...acc, symbol, ...members];
  }, []);
}

export function renderClass(
  symbol: Symbol,
  aliasedSymbol: Symbol,
  context: RenderContext
): string {
  const name = symbol.getName();

  return joinSections([
    heading(name, 2),
    renderDescription(
      aliasedSymbol.getDocumentationComment(context.typeChecker)
    ),
    renderExamples(aliasedSymbol.getJsDocTags()),
    renderAdditionalLinks(aliasedSymbol.getJsDocTags()),
  ]);
}
