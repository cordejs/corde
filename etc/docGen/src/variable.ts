import { Symbol, Type } from 'typescript';
import { renderAdditionalLinks } from './additionalLinks';
import { RenderContext } from './context';
import { renderDescription } from './description';
import { renderExamples } from './examples';
import { heading, joinSections, subSection } from './markdown';
import { renderType } from './type';

export function renderVariable(
  symbol: Symbol,
  aliasedSymbol: Symbol,
  type: Type,
  context: RenderContext
): string {
  return joinSections([
    heading(symbol.getName(), 2),
    renderDescription(
      aliasedSymbol.getDocumentationComment(context.typeChecker)
    ),
    subSection('Type'),
    renderType(type, context),
    renderExamples(aliasedSymbol.getJsDocTags()),
    renderAdditionalLinks(aliasedSymbol.getJsDocTags()),
  ]);
}
