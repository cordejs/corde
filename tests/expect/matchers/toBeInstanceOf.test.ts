import { cordeExpect } from "../../../src/expect";

class Class {}

describe("testing toBeInstanceOf", () => {
  it.each([
    [new Date(), Date],
    [new String(), String],
    [new Number(), Number],
    [new Boolean(), Boolean],
    [new Class(), Class],
    [new Function(), Function],
    [{}, Object],
    [[], Array],
    [new Map(), Map],
    [cordeExpect.any(), Map],
    [cordeExpect.any(String), cordeExpect.any()],
    [cordeExpect.any(), cordeExpect.any(String)],
    [cordeExpect.any(String), cordeExpect.any(String)],
  ])("should return true for %s (isNot false)", (value, type) => {
    expect(cordeExpect(value).toBeInstanceOf(type)).toEqual({ pass: true, message: "" });
  });

  it.each([
    [new Date(), Date],
    [new String(), String],
    [new Number(), Number],
    [new Boolean(), Boolean],
    [new Class(), Class],
    [new Function(), Function],
    [{}, Object],
    [[], Array],
    [new Map(), Map],
  ])("should return false for %s (isNot true)", (value, type) => {
    const report = cordeExpect(value).not.toBeInstanceOf(type);
    expect(report.pass).toBeFalsy();
    expect(report.message).toMatchSnapshot();
  });

  it.each([
    [{}, String],
    [Symbol.for(""), Symbol],
    [cordeExpect.any(Number), String],
    [null, null],
    [undefined, undefined],
  ])("should return false for %s (isNot false)", (value, type) => {
    const report = cordeExpect(value).toBeInstanceOf(type);
    expect(report.pass).toBeFalsy();
    expect(report.message).toMatchSnapshot();
  });
});
