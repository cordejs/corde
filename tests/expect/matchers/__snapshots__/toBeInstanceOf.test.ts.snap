// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing toBeInstanceOf should return false for [] (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mArray[39m.
    got: [31mArray[39m
"
`;

exports[`testing toBeInstanceOf should return false for [Boolean: false] (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mBoolean[39m.
    got: [31mBoolean[39m
"
`;

exports[`testing toBeInstanceOf should return false for [Number: 0] (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mNumber[39m.
    got: [31mNumber[39m
"
`;

exports[`testing toBeInstanceOf should return false for [String: ''] (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mString[39m.
    got: [31mString[39m
"
`;

exports[`testing toBeInstanceOf should return false for {} (isNot false) 1`] = `
"    
    expect([32mexpected[39m).toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should be instance of [32mString[39m.
    got: [31mObject[39m
"
`;

exports[`testing toBeInstanceOf should return false for {} (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mObject[39m.
    got: [31mObject[39m
"
`;

exports[`testing toBeInstanceOf should return false for 2021-08-22T03:18:41.474Z (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mDate[39m.
    got: [31mDate[39m
"
`;

exports[`testing toBeInstanceOf should return false for Class {} (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mClass[39m.
    got: [31mClass[39m
"
`;

exports[`testing toBeInstanceOf should return false for Map(0) {} (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mMap[39m.
    got: [31mMap[39m
"
`;

exports[`testing toBeInstanceOf should return false for Symbol() (isNot false) 1`] = `
"    
    expect([32mexpected[39m).toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should be instance of [32mSymbol[39m.
    got: [31mSymbol[39m
"
`;

exports[`testing toBeInstanceOf should return false for any(Number) (isNot false) 1`] = `
"    
    expect([32mexpected[39m).toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should be instance of [32mString[39m.
    got: [31many(Number)[39m
"
`;

exports[`testing toBeInstanceOf should return false for function anonymous(
) {

} (isNot true) 1`] = `
"    
    expect([32mexpected[39m).not.toBeInstanceOf([31minstanceType[39m)
    
    [1mexpected[22m should not be instance of [32mFunction[39m.
    got: [31mFunction[39m
"
`;

exports[`testing toBeInstanceOf should return false for null (isNot false) 1`] = `"Right-hand side of 'instanceof' is not an object"`;

exports[`testing toBeInstanceOf should return false for undefined (isNot false) 1`] = `"Right-hand side of 'instanceof' is not an object"`;
